### PR TITLE
Merging Active Sensor and DDA Cloud Coefficients into V3

### LINKED ISSUE #44: Add active sensor module to CRTMv3
Add active sensor module for radar and lidar applications to CRTMv3 as well.

### LINKED ISSUE #45: Add DDA hydrometeor LUT to CRTMv3
Add DDA hydrometeor LUT for cloud scattering properties to CRTMv3 as well (this was already available in the scalar release v2.4.1).

### PR DESCRIPTION
**Pull request by imoradi**
## Description

This will merge active sensor module and also the code changes to be able to run CRTM using the DDA hydrometeor lookup tables.

## Issue(s) addressed

- Resolves #44 
- Resolves #45 

## Dependencies

List the other PRs that this PR is dependent on:

I have already merged the changes from other PR #37 into this PR.

## Impact

This should make CRTM V3.0 able to compute reflectivities and work with the DDA lookup tables.

## Checklist

- I have run this version for both active sensor to compute reflectivities and also run new cloud lookup tables. I have also compared the computed reflectivities from V2.4.1 and they are identical.

### CODE REVIEW

**Comment by StegmannJCSDA** [comment]
@imoradi testing now.

**Comment by StegmannJCSDA** [comment]
`test_AD_Active_Sensor.f90` fails to build with the following message from gfortran:
```Fortran
CRTMv3/test/mains/unit/Unit_Test/test_AD_Active_Sensor.f90:405:4:

  405 |     RTSolution_AD%Reflectivity_Attenuated(jj) = ZERO
      |    1
Error: Component to the right of a part reference with nonzero rank must not have the ALLOCATABLE attribute at (1)
CRTMv3/test/mains/unit/Unit_Test/test_AD_Active_Sensor.f90:406:4:

  406 |     RTSolution_AD%Reflectivity(jj) = ZERO
      |    1
Error: Component to the right of a part reference with nonzero rank must not have the ALLOCATABLE attribute at (1)
CRTMv3/test/mains/unit/Unit_Test/test_AD_Active_Sensor.f90:408:7:

  408 |        RTSolution_AD(ichan,:)%Reflectivity_Attenuated(jj) = ONE
      |       1
Error: Component to the right of a part reference with nonzero rank must not have the ALLOCATABLE attribute at (1)
CRTMv3/test/mains/unit/Unit_Test/test_AD_Active_Sensor.f90:410:7:

  410 |        RTSolution_AD(ichan,:)%Reflectivity(jj) = ONE
      |       1
Error: Component to the right of a part reference with nonzero rank must not have the ALLOCATABLE attribute at (1)
```

**Comment by imoradi** [comment]
Are you testing it on S4? If so then can you share with me your gfort.setup

**Comment by StegmannJCSDA** [comment]
> Are you testing it on S4? If so then can you share with me your gfort.setup

No, I am testing locally and I am using the ecbuild build.

**Comment by imoradi** [comment]
Can someone point me to the docs showing how to load ecbuild modules for gfort on S4?

**Comment by imoradi** [comment]
This is probably a gfort issues not allowing to initialize an array using an scalar (ifort is fine). Can you try test_AD_Active_Sensor.f90 to see if it compiles now?

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,11 @@
 .DS_Store
 *.o
+*.rej
```
Comment (StegmannJCSDA) [inline .gitignore]: We shouldn't really have to deal with `*.rej` files.

File: src/Build/CMakeFiles/CMakeOutput.log
Comment (StegmannJCSDA) [inline src/Build/CMakeFiles/CMakeOutput.log:1]: This file needs to be removed.

File: src/Coefficients/CloudCoeff/CloudCoeff_Define.f90 [hunk 2]
```diff
@@ -725,15 +792,28 @@ SUBROUTINE CloudCoeff_Inspect( CloudCoeff, Pause )
...
-      WRITE(*,'(7x,"Density         : ",es13.6)') CloudCoeff%Density(j)
+      WRITE(*,'(7x,"Density_IR         : ",es13.6)') CloudCoeff%Density_IR(j)
       DO i = 1, CloudCoeff%n_IR_Radii
         WRITE(*,'(7x,"Effective radius: ",es13.6)') CloudCoeff%Reff_IR(i)
         WRITE(*,'(5(1x,es13.6,:))') CloudCoeff%g_IR(:,i,j)     
       END DO
     END DO
     
+    IF ( wait ) THEN
+      WRITE(*,'(/5x,"Press <ENTER> to view the infrared solid phase backscatteringparameter")')
```
Comment (StegmannJCSDA) [inline src/Coefficients/CloudCoeff/CloudCoeff_Define.f90]: There's no space in `backscatteringparameter`.

**Review by StegmannJCSDA — CHANGES_REQUESTED** [review]
The PR is very extensive again and the following tests fail:
```
	 41 - test_k_matrix_Simple_atms_n21 (SEGFAULT)
	 42 - test_k_matrix_Simple_cris-fsr_n21 (SEGFAULT)
	 43 - test_k_matrix_Simple_v.abi_g18 (SEGFAULT)
	 44 - test_k_matrix_Simple_modis_aqua (SEGFAULT)
	 72 - test_adjoint_Simple_atms_n21 (Failed)
	 73 - test_adjoint_Simple_cris-fsr_n21 (Failed)
	 74 - test_adjoint_Simple_v.abi_g18 (Failed)
	 75 - test_adjoint_Simple_modis_aqua (Failed)
	 88 - test_Active_Sensor (Failed)
	 90 - test_TL_convergence_active_sensor (Failed)
```
The only ctest that hasn't been fixed yet is `test_Active_Sensor`, which fails because the following file still needs to be added to the testing data before the PR can be merged:
`./testinput/CloudCoeff_DDA_ARTS.nc4`.

File: src/AtmScatter/CRTM_CloudScatter.f90 [hunk 3]
```diff
@@ -317,7 +363,11 @@ FUNCTION CRTM_Compute_CloudScatter( &
         ! compute the single scatter albedo in the Layer_loop below.
         CScat%Optical_Depth(kc) = CScat%Optical_Depth(kc) + &
                                   (CSV%ke(kc,n)*Atm%Cloud(n)%Water_Content(kc))
-
+        IF ( SC(SensorIndex)%Is_Active_Sensor .and. CScat%Include_Scattering ) THEN
```
Comment (BenjaminTJohnson) [inline src/AtmScatter/CRTM_CloudScatter.f90:366]: This use of a flag is inconsistent with:  
```
USE CRTM_SpcCoeff,            ONLY: SC, &
SpcCoeff_IsMicrowaveSensor , & 
SpcCoeff_IsInfraredSensor  , &
SpcCoeff_IsVisibleSensor   , &
SpcCoeff_IsUltravioletSensor
```

File: src/AtmScatter/CRTM_CloudScatter.f90 [hunk 4]
```diff
@@ -508,10 +558,14 @@ FUNCTION CRTM_Compute_CloudScatter_TL( &
           CALL Get_Cloud_Opt_MW_TL(CScat_TL                            , & ! Input
                                    Atm%Cloud(n)%Type                   , & ! Input
                                    CSV%ke(kc,n)                        , & ! Input
+                                   CSV%kb(kc,n)                        , & ! Input
                                    CSV%w(kc,n)                         , & ! Input
+                                   SC(SensorIndex)%Is_Active_Sensor    , & ! Input
```
Comment (BenjaminTJohnson) [inline src/AtmScatter/CRTM_CloudScatter.f90:563]: This should be treated exactly the same as other sensor types, such as `SpcCoeff_IsMicrowaveSensor`

File: src/AtmScatter/CRTM_CloudScatter.f90 [hunk 5]
```diff
@@ -521,25 +575,31 @@ FUNCTION CRTM_Compute_CloudScatter_TL( &
           CALL Get_Cloud_Opt_IR_TL(CScat_TL                            , & ! Input
                                    Atm%Cloud(n)%Type                   , & ! Input
                                    CSV%ke(kc,n)                        , & ! Input
+                                   CSV%kb(kc,n)                        , & ! Input
                                    CSV%w(kc,n)                         , & ! Input
+                                   SC(SensorIndex)%Is_Active_Sensor    , & ! Input
```
Comment (BenjaminTJohnson) [inline src/AtmScatter/CRTM_CloudScatter.f90:580]: I won't comment further, but throughout this needs to be consistent.

File: src/AtmScatter/CRTM_CloudScatter.f90 [hunk 6]
```diff
@@ -1212,65 +1350,61 @@ SUBROUTINE Get_Cloud_Opt_MW( CloudScatter, &  ! Input  CloudScatter structure
...
+               END DO
+             END DO
+
+           ELSE
+           ! Absorption coefficient
+             ke = ke * (ONE - w)
+           END IF
+        END IF ! ICE_Cloud etc interpol
+      CASE DEFAULT
+         print*, 'Cloud State not defined'
```
Comment (BenjaminTJohnson) [inline src/AtmScatter/CRTM_CloudScatter.f90]: please use the built-in message system.

File: src/Atmosphere/CRTM_Atmosphere.f90 [hunk 7]
```diff
@@ -460,6 +472,11 @@ FUNCTION CRTM_Atmosphere_AddLayers_TL( &
     ! Set up
     err_stat = SUCCESS
 
+    IF (.NOT. Atm_In%Add_Extra_Layers) THEN
+        ! This will just copy Atm_In_TL to Atm_Out_TL
+        atm_out_TL = CRTM_Atmosphere_AddLayerCopy( Atm_In_TL, 0 )
```
Comment (BenjaminTJohnson) [inline src/Atmosphere/CRTM_Atmosphere.f90:477]: this looks like a bug fix -- was this something specific that you found?

File: src/Atmosphere/CRTM_Atmosphere_Define.f90 [hunk 8]
```diff
@@ -370,6 +449,7 @@ MODULE CRTM_Atmosphere_Define
     INTEGER, ALLOCATABLE :: Absorber_Units(:) ! J
     ! Profile LEVEL and LAYER quantities
     REAL(fp), ALLOCATABLE :: Level_Pressure(:)  ! 0:K
+    REAL(fp), ALLOCATABLE :: Height(:)          ! 0:K in km
```
Comment (BenjaminTJohnson) [inline src/Atmosphere/CRTM_Atmosphere_Define.f90:452]: Going to need some documentation on this one:  is this height above the surface?   i.e., is the surface pressure level always Height = 0 km?   if so, I'd prefer to name it "Level_Height".

File: src/Build/libsrc/test/check_crtm.F90 [hunk 9]
```diff
@@ -308,8 +308,13 @@ PROGRAM check_crtm
     ! 7b. Inintialize the K-matrix INPUT so
     !     that the results are dTb/dx
     ! -------------------------------------
+    CALL CRTM_RTSolution_Create( rts_K, N_LAYERS ) ! This is necessary for Reflectivity
     rts_K%Radiance               = ZERO
     rts_K%Brightness_Temperature = ONE
+    DO m = 1, N_LAYERS
+       rts_K%Reflectivity(m) = ZERO
```
Comment (BenjaminTJohnson) [inline src/Build/libsrc/test/check_crtm.F90:315]: This seems inelegant compared to the above lines using broadcast to set to zero -- why was this necessary?

File: src/CRTM_LifeCycle.f90 [hunk 10]
```diff
@@ -670,15 +661,20 @@ FUNCTION CRTM_Init( &
 
     ! Load the cloud coefficients
     IF ( Local_Load_CloudCoeff ) THEN
-      ! IF ( iQuiet == 0 ) THEN ! IF ( .not. Quiet) THEN
-      !   WRITE(*, '("Load the cloud coefficients: ") ')
-      !   WRITE(*, '("...Cloud model: ", a) ') TRIM(Default_Cloud_Model)
-      !   WRITE(*, '("...CloudCoeff file: ", a) ') TRIM(Default_CloudCoeff_File)
-      ! END IF
+      IF ( Default_CloudCoeff_Format == 'netCDF' ) THEN
```
Comment (BenjaminTJohnson) [inline src/CRTM_LifeCycle.f90]: @chengdang are you okay with these changes?

File: src/Atmosphere/CRTM_Atmosphere_Define.f90 (hunk 8 above)
Comment (BenjaminTJohnson) [reply src/Atmosphere/CRTM_Atmosphere_Define.f90:452]: Also, where is height used?  I have a concern about level height vs. layer height.

File: src/RTSolution/CRTM_Active_Sensor.f90 [hunk 11]
```diff
@@ -0,0 +1,766 @@
...
+       Atm%Height = Calculate_Height(Atm)
+   END IF
+
+   dZ_m = (Atm%Height(0:n_Layers-1) - Atm%Height(1:n_Layers)) * ONE_THOUSAND
+   IF (PRESENT(GeometryInfo)) THEN
+       dZ_m = dZ_m / GeometryInfo%Cosine_Sensor_Zenith
+   ENDIF
+
+   DO n = 1, Atm%n_Clouds
+      Atm%Cloud(n)%Water_Density = Atm%Cloud(n)%Water_Content / dZ_m
```
Comment (BenjaminTJohnson) [inline src/RTSolution/CRTM_Active_Sensor.f90:153]: because it will result in division by zero here

File: src/RTSolution/CRTM_Active_Sensor.f90 [hunk 12]
```diff
@@ -0,0 +1,766 @@
...
+   ! Calculate heights if hasn't been set already
+   IF (ALL(Atm%Height .LT. EPSILON_FP)) THEN
+       Atm%Height = Calculate_Height(Atm)
+   END IF
+
+   dZ_m = (Atm%Height(0:n_Layers-1) - Atm%Height(1:n_Layers)) * ONE_THOUSAND
+   IF (PRESENT(GeometryInfo)) THEN
+       dZ_m = dZ_m / GeometryInfo%Cosine_Sensor_Zenith
+   ENDIF
+
```
Comment (BenjaminTJohnson) [inline src/RTSolution/CRTM_Active_Sensor.f90:151]: from experience, there needs to be a check to ensure `dZ_m > 0. `

File: src/RTSolution/CRTM_Active_Sensor.f90 [hunk 13]
```diff
@@ -0,0 +1,766 @@
...
+!                       TYPE:       REAL(fp)
+!                       DIMENSION:  1-D Complex Array
+!
+!--------------------------------------------------------------------------------
+
+FUNCTION Water_Permittivity_Turner_2016(freq_Hz, temp_K) RESULT(perm)
+
+  REAL(fp), INTENT(IN) :: freq_Hz, temp_K(:)
+  REAL(fp), DIMENSION(SIZE(temp_K)) :: perm_re, perm_im
+  COMPLEX :: perm(SIZE(temp_K))
```
Comment (BenjaminTJohnson) [inline src/RTSolution/CRTM_Active_Sensor.f90:198]: use `COMPLEX(fp)`

File: src/RTSolution/CRTM_Active_Sensor.f90 [hunk 14]
```diff
@@ -0,0 +1,766 @@
...
+  c1 = 1.302E-13
+  c2 = 1.012E-14
+
+  d1 = 6.627E+2
+  d2 = 6.089E+2
+
+  tc = 133.1383
+
+  e0 = 87.9144
+  e1 = 0.404399_fp
```
Comment (BenjaminTJohnson) [inline src/RTSolution/CRTM_Active_Sensor.f90:227]: why do some constants get `_fp` and others don't?

File: src/RTSolution/CRTM_Active_Sensor.f90 [hunk 15]
```diff
@@ -0,0 +1,766 @@
...
+
+    ! Calculate heights if hasn't been set already
+    IF (ALL(Atm%Height .LT. EPSILON_FP)) THEN
+        Height = Calculate_Height(Atm)
+    ELSE
+        Height = Atm%Height
+    ENDIF
+    dZ_m = (Atm%Height(0:Atm%n_Layers-1) - Atm%Height(1:Atm%n_Layers)) * ONE_THOUSAND
+    dZ_m = dZ_m / GeometryInfo%Cosine_Sensor_Zenith
+
```
Comment (BenjaminTJohnson) [inline src/RTSolution/CRTM_Active_Sensor.f90:355]: again check for `dZ_m > 0`

File: src/RTSolution/CRTM_Active_Sensor.f90 [hunk 16]
```diff
@@ -0,0 +1,766 @@
...
+    ELSE IF( SpcCoeff_IsInfraredSensor(SC(SensorIndex)) ) THEN
+       Wavenumber = SC(SensorIndex)%Wavenumber(ChannelIndex) ! 1/cm
+       Wavelength_m = POINT_01 / Wavenumber
+    END IF
+
+    Temp_K = 273.15_fp
+    perm = Water_Permittivity_Turner_2016(Frequency * 1.0d9, & ! Input
+                                          Temp_K) ! Input
+
+    perm_re = REAL(REAL(perm))  ! Double REAL is required to avoud issues with GNU Fortran
```
Comment (BenjaminTJohnson) [inline src/RTSolution/CRTM_Active_Sensor.f90:368]: what is the issue?  I've never seen this before.

File: src/RTSolution/CRTM_Active_Sensor.f90 [hunk 17]
```diff
@@ -0,0 +1,766 @@
...
+!
+!--------------------------------------------------------------------------------
+
+  SUBROUTINE CRTM_Compute_Reflectivity(Atm         , &  ! Input
+                                       AtmOptics   , &  ! Input
+                                       GeometryInfo , &  ! Input
+                                       SensorIndex , &  ! Input
+                                       ChannelIndex, &  ! Input
+                                       RTSolution )     ! Input/Output
+    ! Arguments
```
Comment (BenjaminTJohnson) [inline src/RTSolution/CRTM_Active_Sensor.f90:329]: all subroutines should start with `implicit none`

File: src/RTSolution/CRTM_Active_Sensor.f90 [hunk 18]
```diff
@@ -0,0 +1,766 @@
...
+    perm = Water_Permittivity_Turner_2016(Frequency * 1.0d9, & ! Input
+                                          Temp_K) ! Input
+
+    perm_re = REAL(REAL(perm))  ! Double REAL is required to avoud issues with GNU Fortran
+    perm_im = REAL(AIMAG(perm))
+    ! perm = cmplx(perm_re, perm_im, FP)
+    kw = (perm - ONE )/(perm + TWO)
+    Kw_2 = ABS(kw)**TWO
+
+    P1 = (M6_MM6 * Wavelength_m**4.0_fp) / (PI**5.0_fp * Kw_2)
```
Comment (BenjaminTJohnson) [inline src/RTSolution/CRTM_Active_Sensor.f90:374]: where is `M6_MM6` defined?

**Review by BenjaminTJohnson — CHANGES_REQUESTED** [review]

**Comment by BenjaminTJohnson** [comment]
@imoradi finished the first round of review, I haven't tested the code yet.

File: src/AtmScatter/CRTM_CloudScatter.f90 (hunk 3 above)
Comment (imoradi) [reply src/AtmScatter/CRTM_CloudScatter.f90:366]: Yes we discussed this when merging into V2.4.1 - basically the active sensors can be MW, IR or anything so we need to see whether a sensor is MW/IR/etc first then active or passive.

File: src/AtmScatter/CRTM_CloudScatter.f90 (hunk 4 above)
Comment (imoradi) [reply src/AtmScatter/CRTM_CloudScatter.f90:563]: Please see my previous comment. Active sensors still need to be one of the other kinds.

File: src/AtmScatter/CRTM_CloudScatter.f90 (hunk 6 above)
Comment (imoradi) [reply src/AtmScatter/CRTM_CloudScatter.f90]: Fixed

File: src/Atmosphere/CRTM_Atmosphere.f90 (hunk 7 above)
Comment (imoradi) [reply src/Atmosphere/CRTM_Atmosphere.f90:477]: In the active module we don't want CRTM to mess with the input layers so I have added this flag to ensure that the output reflectivities are at the same level as the input profiles.

File: src/RTSolution/CRTM_Active_Sensor.f90 (hunk 14 above)
Comment (imoradi) [reply src/RTSolution/CRTM_Active_Sensor.f90:227]: added _fp to other decimals

File: src/RTSolution/CRTM_Active_Sensor.f90 (hunk 15 above)
Comment (imoradi) [reply src/RTSolution/CRTM_Active_Sensor.f90:355]: as long as the pressure is different then dz_m will not be zero and the profiles pressure levels are checked before hands. Why do we want to double check when it should have been checked from the first place? It is easy to put a statement here but just want to make sure it is really necessary.

File: src/RTSolution/CRTM_Active_Sensor.f90 (hunk 16 above)
Comment (imoradi) [reply src/RTSolution/CRTM_Active_Sensor.f90:368]: See this link https://gcc.gnu.org/onlinedocs/gcc-3.4.6/g77/REAL_0028_0029-and-AIMAG_0028_0029-of-Complex.html

File: src/RTSolution/CRTM_Active_Sensor.f90 (hunk 17 above)
Comment (imoradi) [reply src/RTSolution/CRTM_Active_Sensor.f90:329]: I have included IMPLICIT NONE in the top of the module and that should cover all the subroutines. Isn't it? I believe that is the case but if not then will it to each subroutines. I also checked other modules of CRTM and it has only being added once - see also this link https://stackoverflow.com/questions/24337413/where-to-put-implicit-none-in-fortran

File: src/RTSolution/CRTM_Active_Sensor.f90 (hunk 18 above)
Comment (imoradi) [reply src/RTSolution/CRTM_Active_Sensor.f90:374]: In the same file as private variables only accessible to the module subroutines.

**Comment by imoradi** [comment]
> @imoradi finished the first round of review, I haven't tested the code yet.

Thanks @BenjaminTJohnson - I have responded to your comments. Some of the points here were also discussed when merging this into V2.4.1.

**Comment by BenjaminTJohnson** [comment]
In subroutine (`Get_Cloud_Opt_MW`):
```
crtm/src/AtmScatter/CRTM_CloudScatter.f90(1406): error #6404: This name does not have a type, and must have an explicit type.   [ERROR_STATUS]
          Error_Status = FAILURE
----------^
```
It's missing:
``` 
INTEGER :: Error_Status 
<...>
 Error_Status = SUCCESS
<...>
```
You'll also need a variable for the Message:

```
    ! Local variables
    CHARACTER(ML) :: Message
```
and ROUTINE_NAME:

```
    ! Local parameters
    CHARACTER(*), PARAMETER :: ROUTINE_NAME = 'Get_Cloud_Opt_MW'
```

**Comment by BenjaminTJohnson** [comment]
@imoradi This would be much easier to merge if you created a feature branch in the CRTMv3 repository instead of trying to merge from your forked repository, that way I could hotfix your feature branch for minor issues (such as above) to accelerate the process.

**Comment by BenjaminTJohnson** [comment]
@imoradi please merge the latest CRTMv3 develop, and resolve the noted conflicts. 
Thanks!

**Comment by imoradi** [comment]
> @imoradi please merge the latest CRTMv3 develop, and resolve the noted conflicts.

Unfortunately the entire "src/CRTM_Forward_Module.f90" needs once again to be manually checked and merged as changes from the features that were opened after I made this PR are scattered all over this file. I already merged one of the PRs that fixes the aerosol coefficient issue. This was the issue preventing me from using my CRTMv3.0 fork for the JEDI developments.

**Pull request merged**